ouroboros-network/src/Ouroboros/Network/DeltaQ.hs

```diff
@@ -0,0 +1,117 @@
+{-# LANGUAGE NamedFieldPuns             #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE StandaloneDeriving         #-}
+
+module Ouroboros.Network.DeltaQ (
+    -- * GSV type
+    GSV(..),
+    Distribution,
+
+    -- * DeltaQ calculations
+    SizeInBytes,
+    DeltaQ,
+    gsvLeadingEdgeArrive,
+    gsvTrailingEdgeDepart,
+    gsvTrailingEdgeArrive,
+    deltaqQ99thPercentile,
+    deltaqProbabilityMassBeforeDeadline,
+
+    -- * Bi-directional GSV
+    PeerGSV(..),
+    gsvRequestResponseDuration,
+  ) where
+
+import           Data.Semigroup ((<>))
+import           Control.Monad.Class.MonadTimer (TimeMeasure, Duration)
+
+
+--
+-- Sketch of API for GSV and DeltaQ
+--
+
+data Distribution n = Distribution n
+-- This is a totally bogus representation. It's just a PoC.
+-- This says that there's a single value that it takes with probability 1.
+
+instance Num n => Semigroup (Distribution n) where
+  (<>) = convolveDistribution
+
+convolveDistribution :: Num n => Distribution n -> Distribution n -> Distribution n
+convolveDistribution (Distribution d) (Distribution d') = Distribution (d+d')
+-- Again, totally bogus.
+
+shiftDistribution :: Num n => n -> Distribution n -> Distribution n
+shiftDistribution n (Distribution d) = Distribution (n+d)
+-- Again, totally bogus.
+
+data GSV t = GSV (Duration t)                -- G as seconds
+                 (Duration t)                -- S as seconds / octet
+                 (Distribution (Duration t)) -- V as distribution
+
+instance TimeMeasure t => Semigroup (GSV t) where
+  GSV g1 s1 v1 <> GSV g2 s2 v2 = GSV (g1+g2) (s1+s2) (v1 <> v2)
+
+newtype DeltaQ t = DeltaQ (Distribution (Duration t))
+
+deriving instance TimeMeasure t => Semigroup (DeltaQ t)
+
+
+--
+-- Basic calculations based on GSV
+--
+
+type SizeInBytes = Word
+
+gsvLeadingEdgeArrive  :: TimeMeasure t => GSV t ->                DeltaQ t
+gsvTrailingEdgeDepart :: TimeMeasure t => GSV t -> SizeInBytes -> DeltaQ t  -- perhaps a bit dubious
+gsvTrailingEdgeArrive :: TimeMeasure t => GSV t -> SizeInBytes -> DeltaQ t
+
+gsvLeadingEdgeArrive (GSV g _s v) =
+  DeltaQ (shiftDistribution g v)
+
+gsvTrailingEdgeDepart (GSV _g s v) bytes =
+  DeltaQ (shiftDistribution (s * fromIntegral bytes) v)
+
+gsvTrailingEdgeArrive (GSV g s v) bytes =
+  DeltaQ (shiftDistribution (g + s * fromIntegral bytes) v)
```

Comment:
This is possibly the most bogus bit as it assumes ballistic transmission. The next most refined version is one that uses round trips and window opening. Type signatures are fine

Reply:
@njd42 what do you mean by ballistic transmission? Just no congestion control, or something more? And does it have anything to do with :bullettrain_front: (`:bullettrain_front:`) ;)

Reply:
@njd42 so now that S is a function of size, is this now enough for `gsvTrailingEdgeArrive` to capture the non-ballistic cases?